tests/constbuffer_ut/constbuffer_ut.c

```diff
@@ -342,6 +324,101 @@ BEGIN_TEST_SUITE(constbuffer_unittests)
         CONSTBUFFER_Destroy(handle);
     }
 
+    /* CONSTBUFFER_CreateWithMoveMemory */
+
+    /* Tests_SRS_CONSTBUFFER_01_001: [ If source is NULL and size is different than 0 then CONSTBUFFER_CreateWithMoveMemory shall fail and return NULL. ]*/
+    TEST_FUNCTION(CONSTBUFFER_CreateWithMoveMemory_with_invalid_args_fails)
+    {
+        ///arrange
+
+        ///act
+        CONSTBUFFER_HANDLE handle = CONSTBUFFER_CreateWithMoveMemory(NULL, 1);
+
+        ///assert
+        ASSERT_IS_NULL(handle);
+
+        ///cleanup
+    }
+
+    /* Tests_SRS_CONSTBUFFER_01_002: [ Otherwise, CONSTBUFFER_CreateWithMoveMemory shall store the source and size and return a non-NULL handle to the newly created const buffer. ]*/
+    TEST_FUNCTION(CONSTBUFFER_CreateWithMoveMemory_succeeds)
+    {
+        ///arrange
+        CONSTBUFFER_HANDLE handle;
+        const CONSTBUFFER* content;
+        unsigned char* test_buffer = my_gballoc_malloc(2);
+        test_buffer[0] = 42;
+        test_buffer[1] = 43;
+
+        STRICT_EXPECTED_CALL(gballoc_malloc(IGNORED_NUM_ARG));
+
+        ///act
+        handle = CONSTBUFFER_CreateWithMoveMemory(test_buffer, 2);
+
+        ///assert
+        ASSERT_IS_NOT_NULL(handle);
+        /*testing the "storage"*/
+        content = CONSTBUFFER_GetContent(handle);
+        ASSERT_ARE_EQUAL(size_t, 2, content->size);
+        /*testing that it is a copy and not a pointer assignment*/
+        ASSERT_ARE_EQUAL(void_ptr, test_buffer, content->buffer, "same buffer should be returned");
+        ASSERT_ARE_EQUAL(char_ptr, umock_c_get_expected_calls(), umock_c_get_actual_calls());
+
+        ///cleanup
+        CONSTBUFFER_Destroy(handle);
+    }
+
+    /* Tests_SRS_CONSTBUFFER_01_002: [ Otherwise, CONSTBUFFER_CreateWithMoveMemory shall store the source and size and return a non-NULL handle to the newly created const buffer. ]*/
+    TEST_FUNCTION(CONSTBUFFER_CreateWithMoveMemory_with_0_size_succeeds)
+    {
+        ///arrange
+        CONSTBUFFER_HANDLE handle;
+        const CONSTBUFFER* content;
+        unsigned char* test_buffer = my_gballoc_malloc(2);
```

Comment:
>unsigned char* test_buffer = my_gballoc_malloc(2); [](start = 8, length = 50)

I believe you need to free this memory.

Reply:
This is actually on purpose. The buffer passed to this constructor is owned by CONSTBUFFER. It does not matter that it is NULL or non- NULL. malloc can return a non-NULL pointer for a 0 size allocation. Standard says. Imagine this code:

unsigned char* x = malloc(some_size_passed_as_arg);
CONSTBUFFER_HANDLE b = CONSTBUFFER_CreateWithMoveMemory(x, some_size_passed_as_arg);

If some_size_passed_as_arg is 0 malloc can return a non-NULL pointer and then a buffer can be created with that (with size 0).

Well same buffer needs to be freed by someone (the someone that owns it after the Move constructor is definitely the const buffer.